### PR TITLE
RavenDB-21860 improve cluster-wide tx performance

### PR DESCRIPTION
**Pull request by karmeli87**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-21860

### Additional description

- use Http token instead of the operational 15s timeout, this would cause to spam even more requests when the timeout was reached.
- Clone the document outside of the write-tx, since it can be very costly to clone a large document especially if it contain an array with many elements.
- Avoid this cloning when the database execute the transaction
- Avoid this cloning when the leader is getting the command from a follower.
- Avoid leaking token registration callback when we wait for an infinite task.

Tested on our cloud with P10 cluster (sample size is 10)
good-optimizations

### Type of change

- [ ] Bug fix
- [ ] Regression bug fix
- [x] Optimization
- [ ] New feature

### How risky is the change?

- [ ] Low 
- [ ] Moderate 
- [x] High
- [ ] Not relevant

### Backward compatibility

- [ ] Non breaking change
- [x] Ensured. 
The format over the wire as well as the persistent format were not changed.
- [x] Breaking change - minor? 
Cluster tx with not throw on timeout after 15s (operational timeout), but instead are linked with the Http request token. Same as the normal tx.
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [x] No documentation update is needed 

### Testing by Contributor

- [ ] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [x] It has been verified by manual testing

### Testing by RavenDB QA team

- [x] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [ ] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

### CODE REVIEW

File: src/Raven.Server/Documents/DocumentDatabase.cs [hunk 1]
```diff
@@ -389,7 +389,7 @@ record = _serverStore.Cluster.ReadDatabase(context, Name, out index);
                     }
                     catch (Exception e)
                     {
-                        if (_logger.IsInfoEnabled)
+                        if (_logger.IsOperationsEnabled)
                         {
                             _logger.Info("An unhandled exception closed the cluster transaction task", e);
```
Comment (ayende) [inline src/Raven.Server/Documents/DocumentDatabase.cs:394]:
```suggestion
                            _logger.Operations("An unhandled exception closed the cluster transaction task", e);
```

File: src/Raven.Server/Documents/Handlers/BatchHandler.cs [hunk 2]
```diff
@@ -236,8 +237,13 @@ private async Task HandleClusterTransaction(DocumentsOperationContext context, M
             if (topology.Promotables.Contains(ServerStore.NodeTag))
                 throw new DatabaseNotRelevantException("Cluster transaction can't be handled by a promotable node.");
 
-            var clusterTransactionCommand = new ClusterTransactionCommand(Database.Name, Database.IdentityPartsSeparator, topology, command.ParsedCommands, options, raftRequestId);
-            var result = await ServerStore.SendToLeaderAsync(clusterTransactionCommand);
+            using var cts = CreateHttpRequestBoundOperationToken();
+            var clusterTransactionCommand = new ClusterTransactionCommand(Database.Name, Database.IdentityPartsSeparator, topology, command.ParsedCommands, options, raftRequestId)
+            {
+                Timeout = Timeout.InfiniteTimeSpan // we rely on the http token to cancel the command
```
Comment (ayende) [inline src/Raven.Server/Documents/Handlers/BatchHandler.cs:243]: What do you mean here?

File: src/Raven.Server/Documents/Handlers/BatchHandler.cs [hunk 3]
```diff
@@ -248,20 +254,17 @@ private async Task HandleClusterTransaction(DocumentsOperationContext context, M
...
             }
 
-            // wait for the command to be applied on this node
-            await ServerStore.WaitForCommitIndexChange(RachisConsensus.CommitIndexModification.GreaterOrEqual, result.Index);
-
             var array = new DynamicJsonArray();
+            
+            // wait for the command to be applied on this node,
+            // canceling will be done through the cts and not with a timeout
+            await ServerStore.WaitForCommitIndexChange(RachisConsensus.CommitIndexModification.GreaterOrEqual, result.Index, Timeout.InfiniteTimeSpan, cts.Token);
```
Comment (ayende) [inline src/Raven.Server/Documents/Handlers/BatchHandler.cs:261]: The default HTTP timeout is 12 hours, no?

File: src/Raven.Server/Documents/Handlers/BatchHandler.cs (hunk 2 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/Handlers/BatchHandler.cs:243]: basically to have the same behavior as with a single node tx, you can cancel it only when passing a token to the `SaveChanges`.

File: src/Raven.Server/Documents/Handlers/BatchHandler.cs (hunk 3 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/Handlers/BatchHandler.cs:261]: same reason as above

File: src/Raven.Server/ServerWide/ServerStore.cs [hunk 4]
```diff
@@ -3029,20 +3014,21 @@ public DatabaseTopology LoadDatabaseTopology(string databaseName)
             }
         }
 
-        public async Task<(long Index, object Result)> SendToLeaderAsync(CommandBase cmd)
+        // TODO: We need always to link the token with the request abort token
```
Comment (arekpalinski) [inline src/Raven.Server/ServerWide/ServerStore.cs:3017]: TODO?

File: src/Raven.Server/Documents/ClusterTransactionWaiter.cs [hunk 5]
```diff
@@ -42,7 +42,7 @@ public void SetException(string id, long index, Exception e)
             Database.RachisLogIndexNotifications.NotifyListenersAbout(index, e);
             if (_results.TryGetValue(id, out var task))
             {
-                task.SetException(e);
+                task.TrySetException(e);
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/ClusterTransactionWaiter.cs:45]: Why do we use Try... methods now?

Comment (arekpalinski) [reply src/Raven.Server/Documents/ClusterTransactionWaiter.cs:45]: I think I realized why is that - we had this issue that we tried to set the result on an already completed task, right?

File: src/Raven.Server/ServerWide/ServerStore.cs (hunk 4 above)
Comment (karmeli87) [reply src/Raven.Server/ServerWide/ServerStore.cs:3017]: if we accept this PR I'll create a ticket to handle that, didn't want to make changes which aren't directly related to the issue here.

File: src/Raven.Server/Documents/ClusterTransactionWaiter.cs (hunk 5 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/ClusterTransactionWaiter.cs:45]: exactly

**Comment by karmeli87** [comment]
Rebased and opened new PR here https://github.com/ravendb/ravendb/pull/18133

**Pull request closed without merging**